pkg/actors/reminders/reminders.go

```diff
@@ -0,0 +1,975 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reminders
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/utils/clock"
+
+	"github.com/google/uuid"
+
+	"github.com/dapr/components-contrib/state"
+	"github.com/dapr/dapr/pkg/actors/internal"
+	diag "github.com/dapr/dapr/pkg/diagnostics"
+	"github.com/dapr/dapr/pkg/resiliency"
+	"github.com/dapr/kit/logger"
+)
+
+var log = logger.NewLogger("dapr.runtime.actor.reminders")
+
+const (
+	daprSeparator        = "||"
+	metadataPartitionKey = "partitionKey"
+)
+
+// Implements a reminders provider.
+//
+
+type reminders struct {
+	clock                clock.WithTicker
+	runningCh            chan struct{}
+	executeReminderFn    internal.ExecuteReminderFn
+	remindersLock        sync.RWMutex
+	remindersStoringLock sync.Mutex
+	reminders            map[string][]ActorReminderReference
+	activeReminders      *sync.Map
+	evaluationChan       chan struct{}
+	stateStoreProviderFn internal.StateStoreProviderFn
+	resiliency           resiliency.Provider
+	storeName            string
+	evaluationLock       sync.RWMutex
+	config               internal.Config
+	lookUpActorFn        internal.LookupActorFn
+}
+
+// NewRemindersProvider returns a reminders provider.
+func NewRemindersProvider(clock clock.WithTicker, opts internal.RemindersProviderOpts) internal.RemindersProvider {
+	return &reminders{
+		clock:           clock,
+		runningCh:       make(chan struct{}),
+		reminders:       map[string][]ActorReminderReference{},
+		activeReminders: &sync.Map{},
+		evaluationChan:  make(chan struct{}, 1),
+		storeName:       opts.StoreName,
+		config:          opts.Config,
+	}
+}
+
+func (r *reminders) SetExecuteReminderFn(fn internal.ExecuteReminderFn) {
+	r.executeReminderFn = fn
+}
+
+func (r *reminders) SetStateStoreProviderFn(fn internal.StateStoreProviderFn) {
+	r.stateStoreProviderFn = fn
+}
+
+func (r *reminders) SetResiliencyProvider(resiliency resiliency.Provider) {
+	r.resiliency = resiliency
+}
+
+func (r *reminders) SetLookupActorFn(fn internal.LookupActorFn) {
+	r.lookUpActorFn = fn
+}
+
+func (r *reminders) DrainRebalancedReminders(actorType string, actorID string, actorKey string) {
+	r.remindersLock.RLock()
+	reminders := r.reminders[actorType]
+	r.remindersLock.RUnlock()
+
+	for _, rem := range reminders {
+		// r.reminder refers to the actual reminder struct that is saved in the db
```

Comment:
What's r.reminder here?

Reply:
Yup, replaced it with `rem.Reminder`. That should be the new one.